---
title: "Drug Resistance and Susceptibility in Foodborne Illnesses"
subtitle: "Data Exploration"
author: "Breanna Richards, Yanru Liao, Jina Yang"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, error=FALSE, echo = FALSE, fig.width = 6, fig.height = 4)
library(readr)
library(janitor)
library(tidyverse)
library(knitr)
library(kableExtra)
library(stringdist)


```


Link to GitHub Repository: https://github.com/brichards21/PHP2550-Final-Project


## Data Exploration 

The primary data source that we're using for this study is the National Center for Biotechnology Information's (NCBI) Pathogen database. This database includes information about sequenced bacterial pathogens that originate from a food source. It records isolates with source and genetic information. Recall that our study aims to assess if we can predict future isolates' performance against antibiotics. More specifically, can we learn from past isolates to develop a system to recommend antibiotics to new isolates? What kind of patterns can we extract from our current data about behaviors of drug susceptibility and resistance?

The focus of the study started with the Salmonella enterica species. This was partially due to the fact that Salmonella is one of the most common germs that causes foodborne illness in the United States. Due to the nature of the NCBI data, the column containing information about drug resistance and susceptibility is more often then not missing. Since our research question is contingent upon information about how these isolates interact with antibiotics, we only considered cases with non-missing information for this particular variable. Thus, the records that we had available to us to assess our study question were sparse. This led to the decision to extend our analysis outside of Salmonella to also include isolates from the E.coli and Shigella species as well as the Campylobacter jejuni species not only to increase our sample size but to also assess the similarities and differences between the three species. 

Filtering records from all three of the species, our final dataset contained 8,672 records for Salmonella, 3,880 records for E.coli and Shigella, and 3,497 records for Campylobacter jejuni for a total of 16,049 records. We start by examining our variables of interest.


### Variables

There are several variables in the NCBI dataset that we are considering in our study (National Center for Biotechnology Information). 


\textbf{Organism group}

This variable refers to the taxonomy group that the isolate belongs to. For the purpose of our study this is going to take on one of three values: Salmonella enterica, E.coli Shigella, or Campylobacter jejuni.


\textbf{Isolate}

The isolate variable is unique for each record in our data and is how we identify individual organisms that have little genetic mixing. 


\textbf{Strain}

The strain variable gives the microbial strain name. It's used to distinguish a genetically distinct lineage separated from another strain by one or two mutations. Different strings of strains indicate different genetic variants or subtypes of microorganisms. 


\textbf{Collection Date}

Collection date gives the date that the sample was collected.  


\textbf{Location}

The geographical origin of the sample.


\textbf{Source type}

Source type gives the general category that an isolate originates from. Some examples of values of source type are 'Human', 'Animal feed', 'Environmental' and 'Food'. 


\textbf{Host and Host Disease}

The 'Host' variable refers to the host species of the isolate. Examples include Homo sapiens, Bos taurus, pig, chicken, poultry, swine, and bovine. Host disease ties the isolate to a disease origin. Examples include gastroenteritis, salmonellosis, and diarrhea.


\textbf{Serovar}

The serovar gives the distinct variation within a species of bacteria. This column groups isolates based on their surface antigens, allowing the classification of isolates to the subspecies level.


\textbf{Outbreak}

The outbreak variable is a way to group isolates that originated due to the same breakout. It is a submitter-given name for the occurrence of more cases of a disease than expected among a specific group of people or within a specific area over a period of time. 


\textbf{Isolation Source and Isolation Type}

Isolation source describes the physical, environmental and/or local geographical source of the biological sample from which the sampled was derived. Examples include boneless beef, turkey, river, creek water, and water filtration membrane. Isolation type generalizes the isolation source into either clinical or environmental.


\textbf{AST Phenotypes and AMR Genotypes}

The AST Phenotype column gives the antibiotic resistant phenotype of the isolate. This variable is key to our study and lists antibiotics that the isolate has demonstrated resistance, susceptibility, intermediate responsiveness, and/or dose-dependent susceptibility to. AMR genotypes gives the antimicrobial resistant genes found in the isolate.


\textbf{Computed Types}

This formula gives the antigen formula and serotype results from an in-silico experiment. The antigen formula gives the detected presence of specific viral antigen which indicates viral infection. The serotype groups isolates by their distinctive surface structures/antigens. 


\textbf{SNP Cluster}

A single nucleotide polymorphism (SNP) is a genomic variant at a single base position in DNA. It represents the difference in a single DNA building block. A SNP cluster is a group of isolates whose genome assemblies are closely related, depending on the clustering methodology used. This column gives each isolate's pathogen SNP cluster accession.


\textbf{Min-Same and Min Diff}

Min-same is the minimum SNP distance to another isolate of the same isolation type (clinical or environmental). Min-diff is the minimum SNP distance to another isolate of a different isolation type. For example, the minimum SNP difference from an environmental isolate to a clinical isolate.




```{r}


salmonella_isolates <- read_csv("data/raw_data/salmonella_isolates.csv")

ecoli_isolates <- read_csv("data/raw_data/ecoli_isolates.csv")

campylobacter_isolates <- read_csv("data/raw_data/campylobacter_isolates.csv")

isolates_df <- rbind(salmonella_isolates, ecoli_isolates, campylobacter_isolates) %>%
  clean_names() %>%
  rename(organism_group = number_organism_group)

isolates_df <- as.data.frame(isolates_df) %>%
  select(-c(source_type, bio_sample, assembly, create_date, bio_sample,
            isolate_identifiers))



```

------------------

We start by assessing the missingness in our data. 


```{r}

missing_vals <- apply(isolates_df, 2, function(x) sum(is.na(x))/nrow(isolates_df))
missing_vals_count <- apply(isolates_df, 2, function(x) sum(is.na(x)))

missing_table <- cbind(as.data.frame(missing_vals[missing_vals > 0]), as.vector(missing_vals_count[missing_vals_count > 0]))

colnames(missing_table) <- c("% Missing", "Count Missing")


missing_table_k <- missing_table %>%
  arrange(desc(`Count Missing`)) 

```


\textbf{Table 1} gives the variables with missingness and orders them from most missing to least missing. As a rule of thumb, we may want to omit the variables with over 90% missingness as they may not be informative in our analysis. The variables that fit this criteria in our data are `outbreak`, `host disease`, and `host`. We recognize that our study will not be informed by information about an isolate's host disease, host species, and origin from an outbreak. 

```{r tab_missing_data}


kable(missing_table_k, caption = "Isolates Info from Most to Least Missing") %>%
  kable_styling()

```


```{r}


isolates_df <- isolates_df %>%
  select(-c(outbreak, host_disease, host))


```


After removing the high-missingness variables, we note that 54.7% of the isolate records have at least one missing value for the 14 variables that we have left in our data. 

```{r}


missing_rows <- c()

for (i in 1:nrow(isolates_df)) {
row <- isolates_df[i, ]
missing_rows <- c(missing_rows, any(is.na(row)))
}

#sum(missing_rows)/nrow(isolates_df)



```


We advance by assessing the distributions of our variables, starting with \textbf{strain}. There are 15,348 unique strains represented in our dataset, meaning that 95.6% of the isolates in our data are unique. We may want to consider dropping this variable as a result as it will not be too informative to our research problem.


```{r}


#length(unique(isolates_df$strain))
#length(unique(isolates_df$strain))/nrow(isolates_df)

```


Next, we looked at the \textbf{collection date} variable to examine the time periods that our data span. Since the raw values of this variable differ in terms of the specificity of the dates, i.e., some values only record year while others record the exact month and day as well, we'll reduce this variable to only the year for consistency purposes.


```{r}
isolates_df$collection_date <- substr(isolates_df$collection_date, 1, 4)
```


We observe the proportion of years represented within each organism group.



```{r}

prop_year_bact <- prop.table(table(isolates_df$collection_date, isolates_df$organism_group), 2) %>%
  `*`(100) %>% round(2)

prop_year_bact <- as.data.frame(prop_year_bact) %>%
  rename(Organism = Var1) %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  mutate(across(everything(), as.character))

colnames(prop_year_bact) <- c("Year", colnames(prop_year_bact)[-1])

prop_year_bact[, -1] <- matrix(paste0(as.matrix(prop_year_bact[, -1]), '%'), ncol=3)


prop_year_bact <- as.data.frame(prop_year_bact)


```


Interestingly, Salmonella is the only bacteria that retains isolate data for 2002-2008. Data starts to be retained for E.coli and Shigella in 2009 but shows the most records in 2017 and thereafter. Data collection for isolates for Campylobacter started in 2016 and retained records up to and including the year of 2020. As we advance in this project, we'll use this variable to understand trends in our data overtime, more specifically as it pertains to changes in drug resistance and susceptibility.

```{r}

isolates_df <- isolates_df %>%
  mutate(collection_date = as.numeric(collection_date))

```




Next, let's look at the distribution of \textbf{serovar} within each bacteria of interest. Upon further inspection, we see that we only have non-missing serovar information for Salmonella. There are 150 distinct serovars in our Salmonella data. To reduce the number of groupings and simplify, values will only be distinguished as their own group if they represent over 2% of the isolates in the Salmonella data. Remaining serovar categories will be grouped into an 'Other' classification. 


```{r}


serovar_salm <- isolates_df %>%
  group_by(organism_group, serovar) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  filter(organism_group == "Salmonella enterica") %>%
  arrange(desc(freq))

serovar_over1 <- unique(serovar_salm[serovar_salm$freq > 0.02, ]$serovar)

isolates_df <- isolates_df %>%
  mutate(serovar_new = ifelse(serovar %in% serovar_over1, serovar,
                              "Other"))
  
```




```{r}

final_serovar_tbl <- isolates_df[isolates_df$organism_group == "Salmonella enterica", ] %>%
  group_by(serovar_new) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  arrange(desc(freq)) %>%
  rename(Serovar = serovar_new,
         Count = n,
         Freq = freq)


```


With this new sub-classified serovar column, we are left with 13 distinct groupings. Aside from the 'Other' Classification group, Kentucky (~14%), Heidelberg (8%), and Typhimurium var. 5- (7%) are the most common serovar groupings. These groupings may be broad enough to be of interest in our overall research project as we will test their association with drug response, although we note that we can only be able to implement this information in the context of Salmonella due to the constraints of our data.


The \textbf{AST phenotypes} variable, in its rawest form, comes to us as one long string with all of the recorded antibiotics that each isolate is resistant, susceptible, intermediate in response, and susceptible-dose dependent to. We separated and classified this variable into two new variables, collapsing resistant and intermediate drug responses into 'resistant', and collapsing susceptible and susceptible-dose dependent responses into 'susceptible', and listed the corresponding antibiotics for each category. This process was done for ease of use when we eventually go on to build our model and want to separate antibiotics by isolates' responses to them. 





```{r}


AST_df <- isolates_df[, c("isolate", "ast_phenotypes")] 

reorg_AST_df <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(reorg_AST_df) <- c("resist_comb", "suscept_comb",
                            "resist", "suscept")

for (i in 1:nrow(AST_df)) {
  
  R_vec <- c() # resistant
S_vec <- c() # susceptible
R_comb <- c() # resistant & intermediate
S_comb <- c() # susceptible & susceptible-dose dependent


  pheno <- AST_df[i, ]$ast_phenotypes
  split_pheno <- str_split(pheno, ",")
  for (j in 1:length(split_pheno[[1]]))
  {
    if(grepl("=R", split_pheno[[1]][j]))
    {
      R_vec <- c(R_vec, split_pheno[[1]][j])
    } 
    
    if (grepl("=S", split_pheno[[1]][j])) {
      S_vec <- c(S_vec, split_pheno[[1]][j])
    }
    
    if (grepl("=R", split_pheno[[1]][j]) | grepl("=I", split_pheno[[1]][j])){
      R_comb <- c(R_comb, split_pheno[[1]][j])
    } 
    
    if (grepl("=S", split_pheno[[1]][j]) | grepl("=SDD", split_pheno[[1]][j])){
      
      S_comb <- c(S_comb, split_pheno[[1]][j])
    }
  
  }
  
  R_vec <- substr(R_vec,1,nchar(R_vec)-2)
  R_vec <- paste(R_vec, collapse = ", ")
  
  S_vec <- substr(S_vec,1,nchar(S_vec)-2)
  S_vec <- paste(S_vec, collapse = ", ")
  
  R_comb <- substr(R_comb,1,nchar(R_comb)-2)
  R_comb <- paste(R_comb, collapse = ", ")
  
  S_comb <- substr(S_comb,1,nchar(S_comb)-2)
  S_comb <- paste(S_comb, collapse = ", ")
  
  
  
   reorg_AST_df <- rbind(reorg_AST_df, 
                          data.frame("resist_comb" = R_comb,
                                     "suscept_comb" = S_comb,
                                     "resist" = R_vec,
                                     "suscept" = S_vec))
}


```


```{r}

isolates_df <- cbind(isolates_df, reorg_AST_df) %>%
  select(-ast_phenotypes)

```



```{r}

# drug resistance
resist_salm <- isolates_df[isolates_df$organism_group == "Salmonella enterica", ]$resist_comb
resist_salm <- resist_salm[!is.na(resist_salm)]
resist_salm <- unlist(strsplit(resist_salm, ","))

resist_ecoli <- isolates_df[isolates_df$organism_group == "E.coli and Shigella", ]$resist_comb
resist_ecoli <- resist_ecoli[!is.na(resist_ecoli)]
resist_ecoli <- unlist(strsplit(resist_ecoli, ","))

resist_camp <- isolates_df[isolates_df$organism_group == "Campylobacter jejuni", ]$resist_comb
resist_camp <- resist_camp[!is.na(resist_camp)]
resist_camp <- unlist(strsplit(resist_camp, ","))



# drug susceptibility 
sus_salm <- isolates_df[isolates_df$organism_group == "Salmonella enterica", ]$suscept_comb
sus_salm <- resist_salm[!is.na(sus_salm)]
sus_salm <- unlist(strsplit(sus_salm, ","))


sus_ecoli <- isolates_df[isolates_df$organism_group == "E.coli and Shigella", ]$suscept_comb
sus_ecoli <- sus_ecoli[!is.na(sus_ecoli)]
sus_ecoli <- unlist(strsplit(sus_ecoli, ","))


sus_camp <- isolates_df[isolates_df$organism_group == "Campylobacter jejuni", ]$suscept_comb
sus_camp <- sus_camp[!is.na(sus_camp)]
sus_camp <- unlist(strsplit(sus_camp, ","))

```


In Salmonella, tetracycline (22%), streptomycin (16%), sulfisoxazole (12%), ampicillin (11%), and amoxicillin-clavulanic acid (7%) are the top antibiotics that our observed isolates are resistant to.

In E.coli, the top drugs that these isolates are resistant to are tetracycline (14%), ampicillin (11%), streptomycin (7%), sulfisoxazole (7%), and ceftriaxone (6%).

In Campylobacter, tetracycline (48%), ciprofloxacin (21%), nalidixic acid (20%), clindamycin (3%), and erythromycin (2%) are the top drugs that isolates show resistance to. 

In general, these bacteria share some of antibiotics that they've grown most resistant to, but also exhibit a few differences.


```{r}

#sort(prop.table(table(trimws(resist_salm))), decreasing = TRUE)
#sort(prop.table(table(trimws(resist_ecoli))), decreasing = TRUE)
#sort(prop.table(table(trimws(resist_camp))), decreasing = TRUE)


```


Regarding the discussion on drug susceptibility, Salmonella isolates saw the most susceptibility to tetracycline (22%), streptomycin (16%), sulfisoxazole (12%), ampicillin (11%), and amoxicillin-clavulanic acid (7%).

In E.coli, isolates and strains were found to be most susceptible to meropenem (8%), ciprofloxacin (7%), nalidixic acid (7%), azithromycin (7%), and cefoxitin (7%).

Finally, in Campylobacter, the drugs those isolates were most susceptible to were gentamicin (13%), erythromycin (13%), florfenicol (12%), azithromycin (12%), and clindamycin (12%).

Interestingly, we see some overlap in certain bacteria being both susceptible and resistant to the same antibiotics. It'll be interesting to explore this more and see if we can examine potential shifts from susceptible to resistant status and vice versa for these illnesses.

```{r}

#sort(prop.table(table(trimws(sus_salm))), decreasing = TRUE)
#sort(prop.table(table(trimws(sus_ecoli))), decreasing = TRUE)
#sort(prop.table(table(trimws(sus_camp))), decreasing = TRUE)

```


Next, we look at the \textbf{locations} represented in our data. We'll reduce the location variable in our data to just be the countries. The vast majority of the records were collected in the United States with 99% of the Salmonella records hailing from the U.S., 85% of the E.coli and Shigella records from the U.S., and 95% of the Campylobacter jejuni records hailing from the U.S. The Netherlands and Germany are the next most popular countries represented. There are a total of 25 distinct countries represented in our dataset. 


```{r}


isolates_df<- isolates_df %>%
  mutate(country = gsub(":.*$", "", location))


```


```{r}


isolates_df <- isolates_df %>%
  mutate(isolation_source = tolower(isolation_source),
         isolation_source_new = ifelse(grepl("pork", isolation_source) |
                                         grepl("swine", isolation_source) |
                                         grepl("hog", isolation_source) |
                                         grepl("chops", isolation_source) |
                                         grepl("salami", isolation_source),
                                       "pork",
                                       ifelse(grepl("chicken", isolation_source),
                                              "chicken",
                                              ifelse(grepl("beef", isolation_source) |
                                                       grepl("GB", isolation_source) |
                                                       grepl("heifer", isolation_source) |
                                                       grepl("cattle", isolation_source),
                                                     "beef",
                                                     ifelse(grepl("turkey", isolation_source), "turkey", 
                                                            ifelse(grepl("ox", isolation_source), "ox",
                                                                   ifelse(grepl("dairy", isolation_source), "dairy",
                                                                          ifelse(grepl("stool", isolation_source) |
                                                                                   grepl("blood", isolation_source) |
                                                                                   grepl("feces", isolation_source) |
                                                                                   grepl("manure", isolation_source) |
                                                                                   grepl("rect", isolation_source) | grepl("anus", isolation_source), "stool",
                                                                                 ifelse(grepl("blood", isolation_source), "blood",
                                                                                        ifelse(grepl("urine", isolation_source), "urine",
                                                                                               ifelse(grepl("shrimp", isolation_source), "shrimp",
                                                                                                      ifelse(grepl("salmon", isolation_source), "salmon",
                                                                                                             ifelse(grepl("egg", isolation_source), "egg",
                                                                                                                    ifelse(grepl("water", isolation_source) |
                                                                                                                             grepl("drainage", isolation_source), "water",
                                                                                                                           ifelse(is.na(isolation_source), NA, "other/unspecified")))))))))))))))



# isolation_source
# pork = string containing pork or swine or hogs or chops or salami
# chicken = string containing chicken
# beef = string containing beef or GB or heifer or cattle
# turkey = string containing turkey 
# ox = string containing steer or ox
# dairy = string containing dairy
# stool = string containing stool or feces or manure or rect or anus
# blood = string containing blood
# urine = string containing urine 
# shrimp = string containing shrimp 
# salmon = string containing salmon 
# egg = string containing egg
# water = water or drainage

# if not NA, other/under-specified 
# NA = NA

```



```{r}


isolates_df <- isolates_df %>%
  mutate(isolation_source_new = ifelse(isolation_source_new %in% c("egg", "salmon", "shrimp"),
                                       "other/unspecified", isolation_source_new))

```


The \textbf{isolation source} variable originally hosted 180 unique values. Upon further inspection, we found that a lot of these distinctions were due to case sensitiveness as well as many variations of the same general source. For example, 'pork', 'pork chops', 'pork chop', and 'Pork'. To simplify, we grouped the isolates into broader categories based on the content we observed in the isolation source variable. We ultimately reduced this variable to 9 distinct categories: pork, chicken, turkey, beef, stool, urine, water, other/unspecified, and NA (for missingness). 


```{r}


prop_source_bact <- prop.table(table(isolates_df$organism_group, isolates_df$isolation_source_new), 1) %>%
  `*`(100) %>% round(2)


prop_source_bact <- as.data.frame(prop_source_bact) %>%
  rename(Organism = Var1) %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  mutate(across(everything(), as.character))

prop_source_bact[, -1] <- matrix(paste0(as.matrix(prop_source_bact[, -1]), '%'), nrow=3)


prop_source_bact <- as.data.frame(prop_source_bact)
           
```

The most common isolate source in our data is chicken for Campylobacter and Salmonella (\textbf{Table 2}).  For E.coli and Shigella, it's turkey, with chicken being the second most common. 


```{r tab3}

kable(prop_source_bact, caption = "Prop. of Sources Represented Per Bacteria") %>%
  kable_styling()

```

In terms of \textbf{isolation type}, there are 1,405 clinical isolates and 14,644 environmental/other isolates. This brings us to ~91% non-clinical (environmental/other) isolates. There's a difference in the distribution of isolation type in the E.coli bacterium group compared to the other two bacteria (\textbf{Table 3}). We're interested in assessing whether or not similarity in isolation source and isolation type is correlated with similarity and drug response and will explore this further.


```{r}


prop_isol_type <- prop.table(table(isolates_df$organism_group , isolates_df$isolation_type), 1)


prop_isol_type <- prop.table(table(isolates_df$organism_group, isolates_df$isolation_type), 1) %>%
  `*`(100) %>% round(2)


prop_isol_type <- as.data.frame(prop_isol_type) %>%
  rename(Organism = Var1) %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  mutate(across(everything(), as.character))

prop_isol_type[, -1] <- matrix(paste0(as.matrix(prop_isol_type[, -1]), '%'), nrow=3)


prop_isol_type <- as.data.frame(prop_isol_type)



```



```{r tab4}


kable(prop_isol_type, caption = "Prop. of Isolation Type Represented Per Bacteria") %>%
  kable_styling()


```

There are 2,613 \textbf{SNP clusters} represented in our data, 811 clusters for Salmonella, 830 for E.coli and Shigella, and 974 for Campylobacter jejuni.

```{r}


# isolates_df %>%
#   group_by(organism_group) %>%
#   summarise(clusts = length(unique(snp_cluster)))



```


There are 3,085 unique sets of \textbf{AMR genotypes} represented in the data, which we are very interested in dissecting and using to supplement our AST phenotype information in our broader study. These genes will be parsed through during the process of clustering and evaluated for biological relevance in relation to drug response. Overall, mdsA (11%), mdsB (11%), blaEC (5%), and acrF (5%) are the most common antimicrobial resistant genes found in our isolates. Breaking it down by bacteria, within Salmonella, the most common genes are mdsA, mdsB, tet(A), aph(3'')-Ib, and aph(6)-Ib. In E.coli, the top 5 are blaEC, acrF, mdtM, tet(A), and aph(3'')-Ib. Finally, in Campylobacter, the most common genes are tet(O), blaOXA-193, 50S_L22_A103V, gyrA_T86I, and aph(3')-IIIa. We undoubtedly observe heterogeneity between the three species.

```{r}

#length(unique(isolates_df$amr_genotypes))
split_genes <- unlist(str_split(unlist(isolates_df$amr_genotypes), ","))
split_genes_sal <- unlist(str_split(unlist(isolates_df[isolates_df$organism_group == "Salmonella enterica",]$amr_genotypes), ","))

split_genes_ecoli <- unlist(str_split(unlist(isolates_df[isolates_df$organism_group == "E.coli and Shigella",]$amr_genotypes), ","))

split_genes_camp <- unlist(str_split(unlist(isolates_df[isolates_df$organism_group == "Campylobacter jejuni",]$amr_genotypes), ","))
  
#sort(prop.table(table(split_genes)), decreasing = T)

#sort(prop.table(table(split_genes_sal)), decreasing = T)
#sort(prop.table(table(split_genes_ecoli)), decreasing = T)
#sort(prop.table(table(split_genes_camp)), decreasing = T)


```

Next, we looked at the \textbf{min-same} and \textbf{min-diff} variables. We observed that the distributions of minimum SNP distance to another isolate whether that isolate be of the same isolation type or different isolation type, are generally right-skewed (\textbf{Figure 1}). This may speak to the presence of outbreaks, isolates closely related to one another that may even have occurred near the same time. Our E.coli and Shigella records seem to be slightly distinct from the other two bacteria in its distributions as if presents a bit more variance in its min-same and min-diff values.


```{r figdist, fig.cap="Distributions of min-same and min-diff of isolates belonging to each bacteria"}

dist_df <- isolates_df[, c("organism_group", "min_same", "min_diff")] %>%
  pivot_longer(!organism_group, names_to = "type", values_to = "value")

ggplot(dist_df, aes(x = value, col = type)) + 
  geom_histogram(aes(y = ..density..,
                                     fill = type),
                                     alpha = 0.3) +
    geom_density(size = 0.8) + 
  facet_wrap(~ organism_group) +
  labs(fill = "", col = "") +
  theme_bw()

  

```

We created a new count variable to indicate the number of isolates that are "close" to another isolate. We'll define a closely related isolate as an isolate within 7 SNPs of another. If an isolate is within 7 SNPs of another clinical isolate and another environmental isolate, that counts as 2. If an isolate is only 'close' to an isolate of the same type but not close to an isolate of a different type and vice versa, that counts as a value of 1. Thus, only going off of the data that is available to us, this variable will only take on discrete values from 0-2.


```{r}


isolates_df <- isolates_df %>%
  mutate(close_min_diff = ifelse(min_diff <= 7, TRUE, FALSE),
         close_min_same = ifelse(min_same <= 7, TRUE, FALSE),
         num_close = close_min_diff + close_min_same) %>%
  select(-c(close_min_diff, close_min_same))


```


Briefly looking at the distribution of this new count variable, we observed that ~66% of isolates from E.coli and Shigella are not 'close' to any other isolates while this proportion is around 25-26% for Campylobacter and Salmonella. Our Campylobacter jejuni isolates are the most close with ~44% having a min-diff \textit{and} min-same value less than 7. This is followed by Salmonella with 35% of records having a min-diff and min-same value less than 7, and then E.coli with only 9% of records having a min-diff and min-same value less than 7. 


```{r}


prop_close <- prop.table(table(isolates_df$organism_group, isolates_df$num_close), 1)




```



The last variable that we'll look at broadly is \textbf{computed types}. We start by separating computed types into the two pieces of information that the variable provides: antigen formula and serotype. Perhaps the most important piece of information that we want to extract from this variable is the antigen formula because we already have serotype information from the serovar variable. We checked to see if this information serotype and serovar from the two different variables, was comparable or not and found that it was. So, from the computed types variable, we only preserved the antigen formula information. 


```{r}


isolates_df <- isolates_df %>%
separate(computed_types, c("antigen_formula", "serotype"), ",serotype=") %>%
  mutate(antigen_formula = str_remove(antigen_formula, "antigen_formula="))

```


This left us with 126 unique antigen formulas. We reduced this further by looking at the distribution of antigen formulas represented. Upon further investigation, we found that the only isolates that don't have missingness for antigen formula are isolates that belong to Salmonella. Categories representing at least 2% of the Salmonella isolates remained their own distinct columns while categories representing less than 2% were grouped into an 'Other' category. 


```{r}

#length(unique(isolates_df$antigen_formula))

count_antigen <- isolates_df[isolates_df$organism_group == "Salmonella enterica", ] %>%
  group_by(antigen_formula) %>%
  summarise(freq = n() / nrow(isolates_df)) %>%
  arrange(desc(freq))

antigen_over_2 <- count_antigen %>%
  filter(freq >= 0.02)

antigen_over_2 <- antigen_over_2$antigen_formula

```


```{r}

isolates_df <- isolates_df %>%
  mutate(antigen_formula_new = ifelse(antigen_formula %in% antigen_over_2, 
                                      antigen_formula,
                                      "Other"))
  

```


```{r}

# prop.table(table(isolates_df[isolates_df$organism_group == "Salmonella enterica",]$organism_group, 
#                  isolates_df[isolates_df$organism_group == "Salmonella enterica",]$antigen_formula_new), 1)

```

Aside from the "other" category, '8:i:z6' (15%), '4:i:1,2' (13%), and '4:r:1,2' (8%), are the top three most common antigen formulas represented in the Salmonella isolates that we have records on. 



-------------

To start to motivate and inform our broader study, we explored two initial questions.

### Initial Subquestions

The first question that we considered is:

Are isolates within the same SNP cluster susceptible/resistant to the same antibiotics?

In order to assess this, we will compute the similarity scores between the drugs that isolates within the same clusters have. To do this, we'll use the `stringsim` function in the `stringdist` package in R (van der Loo M, 2014). The `stringsim` function computes pairwise string similarities between elements of a vector. It returns values between 0 and 1 (inclusive), corresponding to perfect dissimilarity at a value of 0, and perfect similarity at a value of 1. We used the default method for distance calculation, optimal string alignment (restricted Damerau-Levenshtein distance). We assessed the average similarity values for isolates within the same clusters. These values are broken down by drug resistance and drug susceptibility. 


```{r}
# Within-Cluster Similarity


isolates_df <- isolates_df %>% mutate_all(na_if,"") %>%
  mutate(suscept = tolower(suscept),
         resist = tolower(resist),
         suscept_comb = tolower(suscept_comb),
         resist_comb = tolower(resist_comb)
         )


mean_sim_func <- function(string, clusters) {
# string is variable 
  sim <- c()
   
   purrr::map_dbl(seq_along(string), function(i) {

   
  for (j in 1:length(clusters)) {
   
    if (clusters[j] == clusters[i] & i != j &
        !is.na(clusters[j]) & !is.na(clusters[i]))
   {
    sim <- c(sim, stringdist::stringsim(string[i], string[j]))
   }
    
  }
     
    mean(sim)
    
  })
}
   
# NA value if no other isolate in cluster


 
#isolates_df <- isolates_df %>%
#  mutate(suscept_sim = mean_sim_func(suscept_comb, snp_cluster),
#         resist_sim = mean_sim_func(resist_comb, snp_cluster),
#         )


#write_csv(isolates_df, "data/isolates_df_sim_up.csv")

isolates_df <- read_csv("data/isolates_df_sim_up.csv")
isolates_df <- as.data.frame(isolates_df)

```


```{r}

cluster_stats <- isolates_df %>%
  group_by(snp_cluster, suscept_sim, resist_sim) %>%
  summarise(avg_suscept_sim = mean(suscept_sim),
            avg_resist_sim = mean(resist_sim))


summary_clust_sim <- as.data.frame(do.call(cbind, lapply(cluster_stats, summary))) %>%
  select(-snp_cluster)

summary_clust_sim <- summary_clust_sim[, 3:4]

summary_clust_sim <- head(summary_clust_sim, -1)

```


```{r tab5}

kable(summary_clust_sim,
      caption = "Drug Behavior Similarity between Clusters",
      col.names = c("Susceptible Sim.",
                    "Resistant Sim.")) %>%
  kable_styling()

```


\textbf{Table 4} shows that on average, the drug behaviors of isolates within the same cluster tend to be fairly similar with mean and median values hovering between 0.74-0.86. Despite high averages, there are some clusters that seem to have low similarity based on our data (min cluster average of 0.09 for drug susceptibility and 0.05 for drug resistance). 

We decided to group clusters into two categories, 'close' and 'distant'. Close clusters are defined as clusters with average min-diff or average min-same less than or equal to 10 SNPS. All other clusters are categorized as distant. We looked at the average similarity measures for close and distant clusters and found that in general, clusters that we considered 'close' had high similarity when it came to their isolates' drug responses (\textbf{Table 5}).


```{r}


close_dist_clust <- isolates_df %>%
  group_by(snp_cluster) %>%
  summarise(avg_suscept_sim = mean(suscept_sim),
            avg_resist_sim = mean(resist_sim),
            avg_min_diff = mean(min_diff),
            avg_min_same = mean(min_same)
            ) %>%
  mutate(close_clust = ifelse(avg_min_diff <= 10 | avg_min_same <= 10, "close", "distant")) %>%
  group_by(close_clust) %>%
  summarise(avg_sus = mean(avg_suscept_sim, na.rm = T),
            avg_res = mean(avg_resist_sim, na.rm = T),
            n = n())

close_dist_clust <- close_dist_clust %>%
  filter(close_clust %in% c("close", "distant"))
  



```

```{r tab6}

kable(close_dist_clust,
      caption = "Average drug responses for close and distant clusters with closeness being defined as an average min-diff or min-same SNP distance of < 10 SNPS",
      col.names = c("Cluster Type", "Average Susceptibility",
                    "Average Resistance", 
                    "Num Clusters")) %>%
  kable_styling()


```





```{r}


# isolates_df <- isolates_df %>%
#   mutate(iso_suscept_sim = mean_sim_func(suscept, isolation_source_new),
#          iso_resist_sim = mean_sim_func(resist, isolation_source_new),
#          iso_inter_sim = mean_sim_func(inter, isolation_source_new),
#          iso_sdd_sim = mean_sim_func(suscept_dose_dep, isolation_source_new)
#          )


```






The second question that we considered is: 

Does drug resistance and susceptibility change over time?


```{r}
#head(isolates_df)

# resistance
drug_res <- c(isolates_df$resist_comb)
drug_res <- drug_res[!is.na(drug_res)]

drug_res <- unlist(strsplit(drug_res, split = ","))
drug_res <- trimws(drug_res)
drug_res <- unique(drug_res)

# susceptibility
drug_sus <- c(isolates_df$suscept_comb)
drug_sus <- drug_sus[!is.na(drug_sus)]

drug_sus <- unlist(strsplit(drug_sus, split = ","))
drug_sus <- trimws(drug_sus)
drug_sus <- unique(drug_sus)

```



```{r}

resistance_df <- data.frame(matrix(nrow = nrow(isolates_df),
                                   ncol = length(drug_res)))

colnames(resistance_df) <- drug_res

resistance_df <- cbind(isolate = isolates_df$isolate, resistance_df)

for (i in 1:nrow(isolates_df)) {
  for (j in 1:length(drug_res)) {
  if (grepl(drug_res[j], isolates_df[i, "resist_comb"]))
  {
   resistance_df[i, drug_res[j]] <- 1
  }
    else {
      resistance_df[i, drug_res[j]] <- 0
    }
  }
}

resistance_df <- cbind(strain = isolates_df$strain, organism_group = isolates_df$organism_group,
                       collection_date = isolates_df$collection_date,
                       snp_cluster = isolates_df$snp_cluster, resistance_df)



```



```{r}

#write_csv(resistance_df, "data/resistance_df.csv")

resistance_df <- as.data.frame(read_csv("data/resistance_df.csv"))

#colnames(resistance_df) <- gsub("\\...*","",colnames(resistance_df))
```



```{r}

susceptible_df <- data.frame(matrix(nrow = nrow(isolates_df),
                                   ncol = length(drug_sus)))

colnames(susceptible_df) <- drug_sus

susceptible_df <- cbind(isolate = isolates_df$isolate, susceptible_df)

for (i in 1:nrow(isolates_df)) {
  for (j in 1:length(drug_sus)) {
  if (grepl(drug_sus[j], isolates_df[i, "suscept_comb"]))
  {
   susceptible_df[i, drug_sus[j]] <- 1
  }
    else {
      susceptible_df[i, drug_sus[j]] <- 0
    }
  }
}

susceptible_df <- cbind(strain = isolates_df$strain, organism_group = isolates_df$organism_group,
                       collection_date = isolates_df$collection_date,
                       snp_cluster = isolates_df$snp_cluster, susceptible_df)



```


```{r}

#write_csv(susceptible_df, "data/susceptible_df.csv")

susceptible_df <- as.data.frame(read_csv("data/susceptible_df.csv"))

#colnames(resistance_df) <- gsub("\\...*","",colnames(resistance_df))
```


Since we've established that clusters tend to be very similar in terms of their drug responses, we approached this problem by looking at how the mean drug responses per cluster have changed over time. Collection date/Year is the variable that we used for time. In terms of drug response, an isolate within a cluster was given a value of '1' in either a variable representing drug resistance or susceptibility, if it was susceptible or resistant to that drug. Otherwise, an isolate was given a value of 0. When we group these isolates into their clusters and take each cluster's mean response variable by year, these means will be between 0 and 1 (inclusive) because we're taking means of binary responses.

For illustration purposes, we randomly sampled 5 clusters to show potential changes over time. Let's start with drug resistance to the most popular drug, tetracycline.



```{r}

resistance_sum <- resistance_df %>%
  group_by(snp_cluster, collection_date) %>%
  summarise_if(is.numeric, funs(mean)) %>%
  pivot_longer(!c(snp_cluster, collection_date), names_to = "drug",
               values_to = "resistance")

```



```{r}
set.seed(1999)


subset_clust <- resistance_sum %>%
  filter(drug == "tetracycline") %>%
  group_by(snp_cluster, collection_date) %>%
  summarise(n = n()) %>%
  group_by(snp_cluster) %>%
  summarise(n = n()) %>%
  arrange(desc(n)) %>%
  filter(n > 5, !is.na(snp_cluster)) 

rand_5_clusts <- sample(unique(subset_clust$snp_cluster), 5)
```






```{r}

susceptible_sum <- susceptible_df %>%
  group_by(snp_cluster, collection_date) %>%
  summarise_if(is.numeric, funs(mean)) %>%
  pivot_longer(!c(snp_cluster, collection_date), names_to = "drug",
               values_to = "susceptibility")

```


```{r}
set.seed(1999)


subset_clust_sus <- susceptible_sum %>%
  filter(drug == "tetracycline") %>%
  group_by(snp_cluster, collection_date) %>%
  summarise(n = n()) %>%
  group_by(snp_cluster) %>%
  summarise(n = n()) %>%
  arrange(desc(n)) %>%
  filter(n > 5, !is.na(snp_cluster)) 

rand_5_clusts_sus <- sample(unique(subset_clust_sus$snp_cluster), 5)

```





```{r fig2, fig.cap= "5 clusters' mean resistance to Tetracycline over time"}


 rand5_clusts_df <- resistance_sum[resistance_sum$snp_cluster %in% rand_5_clusts, ]

ggplot(rand5_clusts_df[rand5_clusts_df$drug ==
                        "tetracycline",], aes(x = collection_date, y = resistance,
                           col = snp_cluster,
                           group = snp_cluster)) +
  geom_point(size = 2) +
  geom_line(size = 0.6) +
  labs(col = "cluster", x = "year") +
  theme_bw() +
  geom_hline(yintercept = 0.5, col = "red", linetype = "dashed",
             size = 0.6)



```


Evidently, we observed that in general, isolates that seem to be very similar in terms of their drug responses (i.e. isolates within the same cluster), demonstrate variability in their drug resistance over time (\textbf{Figure 2}). If we draw a line at 0.5 and classify average resistance at or above that line as being drug resistant and responses below that line as not being drug resistant, then we can see that for similar isolates, over time, mean responses have crossed the line between being drug resistant and not.

We did the same thing to observe changes over time for drug susceptibility.


```{r fig3, fig.cap= "5 clusters' mean susceptibility to Tetracycline over time"}


 rand5_clusts_df_sus <- susceptible_sum[susceptible_sum$snp_cluster %in% rand_5_clusts_sus, ]

ggplot(rand5_clusts_df_sus[rand5_clusts_df_sus$drug ==
                        "tetracycline",], aes(x = collection_date, y = susceptibility,
                           col = snp_cluster,
                           group = snp_cluster)) +
  geom_point(size = 2) +
  geom_line(size = 0.6) +
  labs(col = "cluster", x = "year") +
  theme_bw() +
  geom_hline(yintercept = 0.5, col = "red", linetype = "dashed",
             size = 0.6)



```



We observed similar behaviors in drug susceptibility. Even in just a random subset of 5 clusters, we observed that over time, some clusters' mean responses crossed that 0.5 threshold of being susceptible or not being susceptible (\textbf{Figure 3}). Thus, we may infer that drug susceptibility and resistance are dynamic over time and is a complex behavior that we may want to understand and adjust for as we move forward in our study.


## Limitations

We recognize that there are several limitations that our data pose to fully exploring our research question. As aforementioned, by only retaining records from the NCBI database that have non-missingness in drug response, we've limited the isolates and cases that we're able to observe as the drug response variable (`AST_phenotypes`) is more often than not missing. Additionally, this variable is defined by the data submitter, so there could be some error or objectivity involved in that process. The NCBI Pathogen Detection Help Documentation site notes that these data are typically submitted using CLSI or EUCAST guidelines, which change over time. Alternatively, drug response classification is decided by an automated instrument which may infer cutoffs. This could mean that records submitted using an earlier standard/guideline may have different resistance and susceptibility criteria for the same antibiotic compound than it would if using a later standard. Even for the same organism and same isolate, different tests may yield different results. This may bring some nuance into the problem space of what we observe in our data as criteria don't seem to be uniform overtime and vary depending on the different standards used to define drug response.

Additionally, some of the variables that we observe only have information for Salmonella isolates. These variables are antigen formula and serovar/serotype. And so, these variables may only be helpful to refine our future analysis for Salmonella but won't be able to contribute to our other two bacteria of interest. It's also a bit of a nuanced process to assess changes in drug response over time. Strains in our dataset are more unique than not, and it's rare that we observe the same strain twice, especially in different years. To try and remedy this, because of the similarity that we saw between isolates within the same cluster, we approached the problem by assessing each cluster's mean response and imposing a 0.5 threshold on those mean responses to categorize those isolates generally being susceptible or not, and resistant or not. This approach is also taking a strong assumption that if an isolate is susceptible or resistant to an antibiotic, then that would recorded in the data, which reasonably, doesn't always hold true. Additionally, we note that genomic responses to antibiotics are more than likely to change over time as bacteria start to learn more about the antibiotics that are used to treat them, and in response may eventually build a resistance to them. This problem space surrounding treatment is ongoing and constantly in flux due to this. Thus, we realize that what is represented in our data now may or may not be a good scope for what could be represented in data like this for future years.


## Conclusion

The use of antibiotics for the treatment of foodborne illnesses is of growing interest in the field of human health. We have grasped a solid understanding of our data and the distributions of the variables represented in our data. We've learned that variables within the same cluster and variables that are considered 'close' in terms of their SNP distance, tend to have similar responses to certain drugs. Similarity in genomic, clinical, and environmental characteristics is going to define the conversation surrounding this study. By assessing the similarity of our isolates, we may be able to pull significant information about patterns of drug response and how they relate to these values/points of similarity.

Moving forward, we aim to go more in depth with our data and extract patterns and relationships between our covariates of interest as they relate to antibiotic resistance and susceptibility. We will specifically focus more on the genes found in our isolates and the connections that we can draw with antibiotic response in that aspect. The broader goal of this study is to investigate the efficacy of and build a classification model to predict a pathogen's response to an antibiotic (resistant or susceptible). Essentially, we want to approach this as a binary classification problem. This model will be trained to consider the optimal hyper-parameters and covariates for a well-considered bias-variance trade-off, and tested to assess accuracy in prediction. The specifications of the model(s) that we will want to try are not solidified as of now but we are considering logistic regression and multiple machine learning methods for binary classification like random forests. We will also assess how effectively we can cluster antibiotic resistant genes and assess how well those clusters may match and define our observed antibiotic resistance and susceptibility behaviors to treatment.

Overall our models will be assessed using classification and clustering performance metrics like area under the ROC curve (AUC), and silhouette scores.


\newpage 

## References

Edgar, R. C. (2010). Search and clustering orders of magnitude faster than BLAST. Bioinformatics (Oxford, England), 26(19), 2460–2461. https://doi.org/10.1093/bioinformatics/btq461


Hashempour-Baltork, F., Hosseini, H., Shojaee-Aliabadi, S., Torbati, M., Alizadeh, A. M., & Alizadeh, M. (2019). Drug Resistance and the Prevention Strategies in Food Borne Bacteria: An Update Review. Advanced Pharmaceutical Bulletin, 9(3), 335–347. https://doi.org/10.15171/apb.2019.041


Hassani, S., Moosavy, M.-H., Gharajalar, S. N., Khatibi, S. A., Hajibemani, A., & Barabadi, Z. (2022). High prevalence of antibiotic resistance in pathogenic foodborne bacteria isolated from bovine milk. Scientific Reports, 12(1), Article 1. https://doi.org/10.1038/s41598-022-07845-6


Rajaei, M., Moosavy, M.-H., Gharajalar, S. N., & Khatibi, S. A. (2021). Antibiotic resistance in the pathogenic foodborne bacteria isolated from raw kebab and hamburger: Phenotypic and genotypic study. BMC Microbiology, 21(1), 272. https://doi.org/10.1186/s12866-021-02326-8


Ren, Y., Chakraborty, T., Doijad, S., Falgenhauer, L., Falgenhauer, J., Goesmann, A., Hauschild, A.-C., Schwengers, O., & Heider, D. (2022). Prediction of antimicrobial resistance based on whole-genome sequencing and machine learning. Bioinformatics, 38(2), 325–334. https://doi.org/10.1093/bioinformatics/btab681


Van Camp, P.-J., Haslam, D. B., & Porollo, A. (2020). Prediction of Antimicrobial Resistance in Gram-Negative Bacteria From Whole-Genome Sequencing Data. Frontiers in Microbiology, 11, 1013. https://doi.org/10.3389/fmicb.2020.01013


National Center for Biotechnology Information. (n.d.). Home - Pathogen Detection - NCBI. National Center for Biotechnology Information. Retrieved October 23, 2022, from https://www.ncbi.nlm.nih.gov/pathogens/ 


van der Loo M (2014). “The stringdist package for approximate string matching.” _The R Journal_, *6*, 111-122.
<https://CRAN.R-project.org/package=stringdist>.




 
