---
title: "Data Exploration PHP2610 Project"
author: "Breanna Richards"
date: "2022-10-11"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, error=FALSE, echo = FALSE)
library(readr)
library(janitor)
library(tidyverse)
library(knitr)
library(kableExtra)
library(RecordLinkage)
library(stringdist)


```


## Data Exploration 

The primary data source that we're using for this study is the National Center for Biotechnology Information's (NCBI) Pathogen database. This database includes information about sequenced bacterial pathogens that originate from a food source. It records isolates with source and genetic information. Recall that our study aims to assess if we can predict future isolates' performance against antibiotics. More specifically, can we learn from past isolates to develop a system to recommend antibiotics to new isolates?

The focus of the study started with the Salmonella enterica species. This was partially due to the fact that Salmonella is one of the most common germs that causes foodborne illness in the United States. Due to the nature of the NCBI data, the column containing information about drug resistance and susceptibility is more often then not missing. Since our research question is contingent upon information about how these isolates interact with antibiotics, we only consider cases with non-missing information for this particular variable. Thus, the records that we had available to us to assess our study question was sparse. This led to the decision to extend our analysis outside of Salmonella to also include isolates from the E.coli and Shigella species as well as the Campylobacter jejuni species not only to extend our records but to also possible assess the similarities and differences between the three species. 

And so, filtering records from all three of the species, our final dataset contained 8,672 records for Salmonella, 3,880 records for E.coli and Shigella, and 3,497 records for Campylobacter jejuni for a total of 16,049 records. Let's start by examining our variables of interest.


### Variables

There are several variables in the NCBI dataset that we are interested in considering in our study. 


\textbf{Organism group}

This variable refers to the taxonomy group that the isolate belongs to. For the purpose of our study this is going to take on one of three values: Salmonella enterica, E.coli Shigella or Campylobacter jejuni.


\textbf{Isolate}

The isolate variable is unique for each record in our data and is how we identify individual organisms that have little genetic mixing. 


\textbf{Strain}

The strain variable gives the microbial strain name. It's used to distinguish a genetically distinct lineage separated from another strain by one or two mutations. Different strings of strains indicate different genetic variants or subtypes of microorganisms. 


\textbf{Collection Date}

Collection date gives the date that the sample was collected.  


\textbf{Location}

The geographical origin of the sample.


\textbf{Source type}

Source type gives the general category that an isolate originates from. Some examples of values of source type are 'Human', 'Animal feed', 'Environmental' and 'Food'. 


\textbf{Host and Host Disease}

The 'Host' variable refers to the host species of the isolate. Examples include Homo sapiens, Bos taurus, pig, chicken, poultry, swine, and bovine. Host disease ties the isolate to a disease origin. Examples include gastroenteritis, salmonellosis, and diarrhea.


\textbf{Serovar}

The serovar gives the distinct variation within a species of bacteria. This column groups isolates based on their surface antigens, allowing the classification of isolates to the subspecies level.


\textbf{Outbreak}

The outbreak variable is a way to group isolates that originated due to the same breakout. It is a submitter-given name for the occurrence of more cases of a disease than expected among a specific group of people or within a specific area over a period of time. 


\textbf{Isolation Source and Isolation Type}

Isolation source describes the physical, environmental and/or local geographical source of the biological sample from which the sampled was derived. Examples include boneless beef, turkey, river, creek water, and water filtration membrane. Isolation type generalizes the isolation source into either clinical or environmental.


\textbf{AST Phenotypes and AMR Genotypes}

The AST Phenotype column gives the antibiotic resistant phenotype of the isolate. This variable is key to our study and lists antibiotics that the isolate has demonstrated resistance, susceptibility, intermediate responsiveness, and/or dose-dependent susceptibility to. AMR genotypes gives the antimicrobial resistant genes found in the isolate.


\textbf{Computed Types}

This formula gives the antigen formula and serotype results from an in-silico experiment. The antigen formula gives the detected presence of specific viral antigen which indicates viral infection. The serotype groups isolates by their distinctive surface structures/antigens. 


\textbf{SNP Cluster}

A single nucleotide polymorphism (SNP) is a genomic variant at a single base postion in DNA. It represents the difference in a single DNA building block. A SNP cluster is a group of isolates whose genome assemblies are closely related, depending on the clustering methodology used. This column gives each isolate's pathogen SNP cluster accession.


\textbf{Min-Same and Min Diff}

Min-same is the minimum SNP distance to another isolate of the same isolation type (clinical or environmental). Min-diff is the minimum SNP distance to another isolate of a different isolation type. For example, the minimum SNP difference from an environmental isolate to a clinical isolate.




```{r}


salmonella_isolates <- read_csv("data/salmonella_isolates.csv")

ecoli_isolates <- read_csv("data/ecoli_isolates.csv")

campylobacter_isolates <- read_csv("data/campylobacter_isolates.csv")

isolates_df <- rbind(salmonella_isolates, ecoli_isolates, campylobacter_isolates) %>%
  clean_names() %>%
  rename(organism_group = number_organism_group)

isolates_df <- as.data.frame(isolates_df) %>%
  select(-c(source_type, bio_sample, assembly, create_date, bio_sample,
            isolate_identifiers))



```

------------------

We start by assessing the missingness in our data. 


```{r}

missing_vals <- apply(isolates_df, 2, function(x) sum(is.na(x))/nrow(isolates_df))
missing_vals_count <- apply(isolates_df, 2, function(x) sum(is.na(x)))

missing_table <- cbind(as.data.frame(missing_vals[missing_vals > 0]), as.vector(missing_vals_count[missing_vals_count > 0]))

colnames(missing_table) <- c("% Missing", "Count Missing")


missing_table_k <- missing_table %>%
  arrange(desc(`Count Missing`)) 

```


\textbf{Table 1} gives the variables with missingness and orders them from most missing to least missing. As a rule of thumb, we may want to omit the variables with over 90% missingness as they may not be too helpful or informative in our analysis. These variables are outbreak, host disease, and host. We recognize that our study will not be informed by information about an isolate's host disease, host species, and origin from an outbreak. 

```{r tab_missing_data}


kable(missing_table_k, caption = "Isolates Info from Most to Least Missing") %>%
  kable_styling()

```


```{r}


isolates_df <- isolates_df %>%
  select(-c(outbreak, host_disease, host))


```


After removing the high-missingness variables, we note that 54.7% of the isolate records have at least one missing value for the 14 variables that we have left in our data. 

```{r}


missing_rows <- c()

for (i in 1:nrow(isolates_df)) {
row <- isolates_df[i, ]
missing_rows <- c(missing_rows, any(is.na(row)))
}

#sum(missing_rows)/nrow(isolates_df)



```


We advance by assessing the distributions of our variables, starting with \textbf{strain}. There are 15,348 unique strains represented in our dataset, meaning that 95.6% of the isolates in our data are unique. We may want to consider dropping this variable as a result as it will not be too informative to our research problem.


```{r}


#length(unique(isolates_df$strain))
#length(unique(isolates_df$strain))/nrow(isolates_df)

```


Next, we looked at the \textbf{collection date} variable to examine the time periods that our data span. Since the raw values of this variable differ in terms of the specificity of the dates, i.e., some values only record year while others record the exact month and day as well, we'll reduce this variable to only the year for consistentcy purposes.


```{r}
isolates_df$collection_date <- substr(isolates_df$collection_date, 1, 4)
```


We observe the proportion of years represented within each organism group.



```{r}

prop_year_bact <- prop.table(table(isolates_df$collection_date, isolates_df$organism_group), 2) %>%
  `*`(100) %>% round(2)

prop_year_bact <- as.data.frame(prop_year_bact) %>%
  rename(Organism = Var1) %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  mutate(across(everything(), as.character))

colnames(prop_year_bact) <- c("Year", colnames(prop_year_bact)[-1])

prop_year_bact[, -1] <- matrix(paste0(as.matrix(prop_year_bact[, -1]), '%'), ncol=3)


prop_year_bact <- as.data.frame(prop_year_bact)


```


```{r tab2}

kable(prop_year_bact, caption = "Prop. of Years Represented Per Bacteria") %>%
  kable_styling()

```


Interestingly, Salmonella is the only bacteria that retains isolate data for 2002-2008 (\textbf{Table 2}). Data starts to be retained for E.coli and Shigella in 2009 but shows the most records in 2017 and thereafter. Data collection for isolates for Campylobacter started in 2016 and retained records up to and including the year of 2020. As we advance in this project, we'll use this variable to understands trends in our data overtime, more specifically as it pertains to changes in drug resistance and susceptibility.

```{r}

isolates_df <- isolates_df %>%
  mutate(collection_date = as.numeric(collection_date))

```




Next, let's look at the distribution of \textbf{serovar} within each bacteria of interest. Upon further inspection, we see that we only have non-missing serovar information for Salmonella. There are 150 distinct serovars in our Salmonella data. To reduce the number of groupings and simplify, values will only be distinguished as their own group if they represent over 2% of the isolates in the Salmonella data. Remaining serovar categories will be grouped into an 'Other' classification. 


```{r}


serovar_salm <- isolates_df %>%
  group_by(organism_group, serovar) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  filter(organism_group == "Salmonella enterica") %>%
  arrange(desc(freq))

serovar_over1 <- unique(serovar_salm[serovar_salm$freq > 0.02, ]$serovar)

isolates_df <- isolates_df %>%
  mutate(serovar_new = ifelse(serovar %in% serovar_over1, serovar,
                              "Other"))
  
```




```{r}

final_serovar_tbl <- isolates_df[isolates_df$organism_group == "Salmonella enterica", ] %>%
  group_by(serovar_new) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  arrange(desc(freq)) %>%
  rename(Serovar = serovar_new,
         Count = n,
         Freq = freq)


```


With this new subclassified serovar column, we are left with 13 distinct groupings. Aside from the 'Other' Classification group, Kentucky (~14%), Heidelberg (8%), and Typhimurium var. 5- (7%) are the most common serovar groupings. 


The \textbf{AST phenotypes} variable, in its rawest form, come to us as one long string with all of the recorded antibiotics that each isolate is resistant, susceptible, intermediate in response, and susceptible-dose dependent to. We separated and classified this variable into four new variables (one for each of the drug responses mentioned) and listed the corresponding antibiotics for each category. This process was done for ease of use when we eventually go on to build our model and want to separate antibiotics by isolates' responses to them. 





```{r}


AST_df <- isolates_df[, c("isolate", "ast_phenotypes")] 

reorg_AST_df <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(reorg_AST_df) <- c("resist", "suscept", "inter", "suscept_dose_dep")

for (i in 1:nrow(AST_df)) {
  
  R_vec <- c() # resistant
S_vec <- c() # susceptible
I_vec <- c() # intermediate
SDD_vec <- c() # susceptible-dose dependent


  pheno <- AST_df[i, ]$ast_phenotypes
  split_pheno <- str_split(pheno, ",")
  for (j in 1:length(split_pheno[[1]]))
  {
    if(grepl("=R", split_pheno[[1]][j]))
    {
      R_vec <- c(R_vec, split_pheno[[1]][j])
    } else if (grepl("=S", split_pheno[[1]][j])) {
      S_vec <- c(S_vec, split_pheno[[1]][j])
    }
    else if (grepl("=I", split_pheno[[1]][j])){
      I_vec <- c(I_vec, split_pheno[[1]][j])
    } else {
      SDD_vec <- c(SDD_vec, split_pheno[[1]][j])
    }
  
  }
  
  R_vec <- substr(R_vec,1,nchar(R_vec)-2)
  R_vec <- paste(R_vec, collapse = ", ")
  
  S_vec <- substr(S_vec,1,nchar(S_vec)-2)
  S_vec <- paste(S_vec, collapse = ", ")
  
  I_vec <- substr(I_vec,1,nchar(I_vec)-2)
  I_vec <- paste(I_vec, collapse = ", ")
  
  SDD_vec <- substr(SDD_vec,1,nchar(SDD_vec)-2)
  SDD_vec <- paste(SDD_vec, collapse = ", ")
  
  
  
   reorg_AST_df <- rbind(reorg_AST_df, 
                          data.frame("resist" = R_vec,
                                     "suscept" = S_vec,
                                     "inter" = I_vec,
                                     "suscept_dose_dep" = SDD_vec))
}



```


```{r}

isolates_df <- cbind(isolates_df, reorg_AST_df) %>%
  select(-ast_phenotypes)

```



```{r}

# drug resistance
resist_salm <- isolates_df[isolates_df$organism_group == "Salmonella enterica", ]$resist
resist_salm <- resist_salm[!is.na(resist_salm)]
resist_salm <- unlist(strsplit(resist_salm, ","))

resist_ecoli <- isolates_df[isolates_df$organism_group == "E.coli and Shigella", ]$resist
resist_ecoli <- resist_ecoli[!is.na(resist_ecoli)]
resist_ecoli <- unlist(strsplit(resist_ecoli, ","))

resist_camp <- isolates_df[isolates_df$organism_group == "Campylobacter jejuni", ]$resist
resist_camp <- resist_camp[!is.na(resist_camp)]
resist_camp <- unlist(strsplit(resist_camp, ","))



# drug susceptibility 
sus_salm <- isolates_df[isolates_df$organism_group == "Salmonella enterica", ]$suscept
sus_salm <- resist_salm[!is.na(sus_salm)]
sus_salm <- unlist(strsplit(sus_salm, ","))


sus_ecoli <- isolates_df[isolates_df$organism_group == "E.coli and Shigella", ]$suscept
sus_ecoli <- sus_ecoli[!is.na(sus_ecoli)]
sus_ecoli <- unlist(strsplit(sus_ecoli, ","))


sus_camp <- isolates_df[isolates_df$organism_group == "Campylobacter jejuni", ]$suscept
sus_camp <- sus_camp[!is.na(sus_camp)]
sus_camp <- unlist(strsplit(sus_camp, ","))

```


In Salmonella, tetracycline (23%), streptomycin (17%), sulfisoxazole (13%), ampicillin (12%), and ceftriaxone (6%) are the top antibiotics that our observed isolates are resistant to.

In E.coli, the top drugs that these isolates are resistant to are tetracycline (15%), ampicilin (12%), streptomycin (8%), sulfisoxazole (8%), and ceftriaxone (7%).

In Campylobacter, tetracycline (48%), ciprofloxacin (21%), nalidixic acid (20%), clindamycin (3%), and erythromycin (2%) are the top drugs that isolates show resistance to. 

In general, these bacteria share a lot of antibiotics that they've grown resistant to.


```{r}

#sort(prop.table(table(trimws(resist_salm))), decreasing = TRUE)
#sort(prop.table(table(trimws(resist_ecoli))), decreasing = TRUE)
#sort(prop.table(table(trimws(resist_camp))), decreasing = TRUE)


```


Regarding the discussion on drug susceptibility, Salmonella isolates saw the most susceptibility to tetracycline (23%), streptomycin (17%), sulfisoxazole (13%), and ampicillin (12%).

In E.coli, isolates and strains were found to be most susceptible to meropenem (8%), ciprofloxacin (7%), nalidixic acid (7%), azithromycin (7%), and cefoxitin (7%).

Finally, in Campylobacter, the drugs those isolates were most susceptible to were gentamicin (13%), erythromycin (13%), florfenicol (13%), azithromycin (12%), and clindamycin (12%).

Interestingly, we see some overlap in certain bacteria being both susceptible and resistant to the same antibiotics. It'll be interesting to explore this more and see if examine potential shifts from susceptible to resistant status for these illnesses.

```{r}

#sort(prop.table(table(trimws(sus_salm))), decreasing = TRUE)
#sort(prop.table(table(trimws(sus_ecoli))), decreasing = TRUE)
#sort(prop.table(table(trimws(sus_camp))), decreasing = TRUE)

```


Next, we look at the \textbf{locations} represented in our data. We'll reduce the location variable in our data to just be the countries. The vast majority of the records were collected in the United States with 99% of the Salmonella records hailing from the U.S., 85% of the E.coli and Shigella records from the U.S., and 95% of the Campylobacter jejuni records hailing from the U.S. The Netherlands and Germany are the next most popular countries represented. There are a total of 25 distinct countries represented in our dataset. 


```{r}


isolates_df<- isolates_df %>%
  mutate(country = gsub(":.*$", "", location))


```


```{r}


isolates_df <- isolates_df %>%
  mutate(isolation_source = tolower(isolation_source),
         isolation_source_new = ifelse(grepl("pork", isolation_source) |
                                         grepl("swine", isolation_source) |
                                         grepl("hog", isolation_source) |
                                         grepl("chops", isolation_source) |
                                         grepl("salami", isolation_source),
                                       "pork",
                                       ifelse(grepl("chicken", isolation_source),
                                              "chicken",
                                              ifelse(grepl("beef", isolation_source) |
                                                       grepl("GB", isolation_source) |
                                                       grepl("heifer", isolation_source) |
                                                       grepl("cattle", isolation_source),
                                                     "beef",
                                                     ifelse(grepl("turkey", isolation_source), "turkey", 
                                                            ifelse(grepl("ox", isolation_source), "ox",
                                                                   ifelse(grepl("dairy", isolation_source), "dairy",
                                                                          ifelse(grepl("stool", isolation_source) |
                                                                                   grepl("blood", isolation_source) |
                                                                                   grepl("feces", isolation_source) |
                                                                                   grepl("manure", isolation_source) |
                                                                                   grepl("rect", isolation_source) | grepl("anus", isolation_source), "stool",
                                                                                 ifelse(grepl("blood", isolation_source), "blood",
                                                                                        ifelse(grepl("urine", isolation_source), "urine",
                                                                                               ifelse(grepl("shrimp", isolation_source), "shrimp",
                                                                                                      ifelse(grepl("salmon", isolation_source), "salmon",
                                                                                                             ifelse(grepl("egg", isolation_source), "egg",
                                                                                                                    ifelse(grepl("water", isolation_source) |
                                                                                                                             grepl("drainage", isolation_source), "water",
                                                                                                                           ifelse(is.na(isolation_source), NA, "other/unspecified")))))))))))))))



# isolation_source
# pork = string containing pork or swine or hogs or chops or salami
# chicken = string containing chicken
# beef = string containing beef or GB or heifer or cattle
# turkey = string containing turkey 
# ox = string containing steer or ox
# dairy = string containing dairy
# stool = string containing stool or fesces or manure or rect or anus
# blood = string containing blood
# urine = string containing urine 
# shrimp = string containing shrimp 
# salmon = string containing salmon 
# egg = string containing egg
# water = water or drainage

# if not NA, other/underspecified 
# NA = NA

```



```{r}


isolates_df <- isolates_df %>%
  mutate(isolation_source_new = ifelse(isolation_source_new %in% c("egg", "salmon", "shrimp"),
                                       "other/unspecified", isolation_source_new))

```


The \textbf{isolation source} variable originally hosted 180 unique values. Upon further inspection, we found that a lot of these distinctions were due to case sensitiveness as well as many variations of the same general source. For example, 'pork', 'pork chops', 'pork chop', and 'Pork'. To simply, we grouped the isolates into broader categories based on the content we observed in the isolation source variable. We ultimately reduced this variable to 9 distinct catgeories: pork, chicken, turkey, beef, stool, urine, water, other/unspecified, and NA (for missingness). 


```{r}


prop_source_bact <- prop.table(table(isolates_df$organism_group, isolates_df$isolation_source_new), 1) %>%
  `*`(100) %>% round(2)


prop_source_bact <- as.data.frame(prop_source_bact) %>%
  rename(Organism = Var1) %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  mutate(across(everything(), as.character))

prop_source_bact[, -1] <- matrix(paste0(as.matrix(prop_source_bact[, -1]), '%'), nrow=3)


prop_source_bact <- as.data.frame(prop_source_bact)
           
```

The most common isolate source in our data is chicken for Campylobacter and Salmonella (\textbf{Table 3}).  For E.coli and Shigella, it's turkey, with chicken being the second most common. 


```{r tab3}

kable(prop_source_bact, caption = "Prop. of Sources Represented Per Bacteria") %>%
  kable_styling()

```

In terms of \textbf{isolation type}, there are 1,405 clinical isolates and 14,644 environmental/other isolates. This brings us to ~91% non-clinical (environmental/other) isolates. There's a difference in the distribution of isolation type in the E.coli bacterium group compared to the other two bacteria (Table 4).


```{r}


prop_isol_type <- prop.table(table(isolates_df$organism_group , isolates_df$isolation_type), 1)


prop_isol_type <- prop.table(table(isolates_df$organism_group, isolates_df$isolation_type), 1) %>%
  `*`(100) %>% round(2)


prop_isol_type <- as.data.frame(prop_isol_type) %>%
  rename(Organism = Var1) %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  mutate(across(everything(), as.character))

prop_isol_type[, -1] <- matrix(paste0(as.matrix(prop_isol_type[, -1]), '%'), nrow=3)


prop_isol_type <- as.data.frame(prop_isol_type)



```



```{r tab4}


kable(prop_isol_type, caption = "Prop. of Isolation Type Represented Per Bacteria") %>%
  kable_styling()


```

There are 2,613 \textbf{SNP clusters} represented in our data, 811 clusters for Salmonella, 830 for E.coli and Shigella, and 974 for Campylobacter jejuni.

```{r}


# isolates_df %>%
#   group_by(organism_group) %>%
#   summarise(clusts = length(unique(snp_cluster)))



```


There are 3,085 unique sets of \textbf{AMR genotypes} represented in the data, which we may be interested in dissecting and using to supplement our AST phenotype information.

```{r}

#length(unique(isolates_df$amr_genotypes))

```

Next, we looked at the \textbf{min-same} and \textbf{min-diff} variables.

```{r figdist, fig.cap="Distributions of Min Same and Min Diff bh Bacteria"}

dist_df <- isolates_df[, c("organism_group", "min_same", "min_diff")] %>%
  pivot_longer(!organism_group, names_to = "type", values_to = "value")

ggplot(dist_df, aes(x = value, col = type)) + 
  geom_histogram(aes(y = ..density..,
                                     fill = type),
                                     alpha = 0.3) +
    geom_density(size = 0.8) + 
  facet_wrap(~ organism_group) +
  labs(fill = "", col = "") +
  theme_bw()

  

```


To Do:


## computed types



-------------

### Initial Subquestions:

The first question that we want to consider is:
1. Are isolates within the same SNP cluster susceptible/resistant to the same antibiotics?

In order to assess this, we will compute the similarity scores between the drugs that isolates within the same clusters have. To do this, we'll use the `stringsim` function in the `stringsim` package in R.


```{r}
# Within-Cluster Similarity


isolates_df <- isolates_df %>% mutate_all(na_if,"") %>%
  mutate(suscept = tolower(suscept),
         resist = tolower(resist),
         inter = tolower(inter),
         suscept_dose_dep = tolower(suscept_dose_dep)
         )


mean_sim_func <- function(string, clusters) {
# string is variable 
  sim <- c()
   
   purrr::map_dbl(seq_along(string), function(i) {

   
  for (j in 1:length(clusters)) {
   
    if (clusters[j] == clusters[i] & i != j &
        !is.na(clusters[j]) & !is.na(clusters[i]))
   {
    sim <- c(sim, stringdist::stringsim(string[i], string[j]))
   }
    
  }
     
    mean(sim)
    
  })
}
   
# NA value if no other isolate in cluster


 
# isolates_df <- isolates_df %>%
#   mutate(suscept_sim = mean_sim_func(suscept, snp_cluster),
#          resist_sim = mean_sim_func(resist, snp_cluster),
#          inter_sim = mean_sim_func(inter, snp_cluster),
#          sdd_sim = mean_sim_func(suscept_dose_dep, snp_cluster)
#          )


#write_csv(isolates_df, "data/isolates_df_sim_up.csv")

isolates_df <- read_csv("data/isolates_df_sim_up.csv")
isolates_df <- as.data.frame(isolates_df)

```



```{r}

# Within Bacteria Similarity 

mean_sim_bact_func <- function(string, org) {
# string is variable 
  sim <- c()
   
   purrr::map_dbl(seq_along(string), function(i) {

   
  for (j in 1:length(org)) {
   
    if (org[j] == org[i] & i != j &
        !is.na(org[j]) & !is.na(org[i]))
   {
    sim <- c(sim, stringdist::stringsim(string[i], string[j]))
   }
    
  }
     
    mean(sim)
    
  })
}


# isolates_df <- isolates_df %>%
#   mutate(bact_suscept_sim = mean_sim_func(suscept, organism_group),
#          bact_resist_sim = mean_sim_func(resist, organism_group),
#          bact_inter_sim = mean_sim_func(inter, organism_group),
#          bact_sdd_sim = mean_sim_func(suscept_dose_dep, organism_group)
#          )


```






The second question that we want to consider is:
How does drug resistance and susceptibility change over time?


- dataset just with isolate, year, and column for each drug in drug resistance - 1 if resistant 0 otherwise

- do the same for susceptible 



\newpage 

References


[3] https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2871933/
